README.md
# Knet

[![](https://img.shields.io/badge/docs-latest-blue.svg)](https://denizyuret.github.io/Knet.jl/latest) 
[![](https://travis-ci.org/denizyuret/Knet.jl.svg?branch=master)](https://travis-ci.org/denizyuret/Knet.jl) 
[![](https://coveralls.io/repos/github/denizyuret/Knet.jl/badge.svg?branch=master)](https://coveralls.io/github/denizyuret/Knet.jl?branch=master)
[![](https://codecov.io/gh/denizyuret/Knet.jl/branch/master/graph/badge.svg)](https://codecov.io/gh/denizyuret/Knet.jl)

[Knet](https://denizyuret.github.io/Knet.jl/latest) (pronounced "kay-net") is the [Koç
University](http://www.ku.edu.tr/en) deep learning framework implemented in
[Julia](http://docs.julialang.org) by [Deniz Yuret](http://www.denizyuret.com) and
collaborators.  It supports GPU operation and automatic differentiation using dynamic
computational graphs for models defined in plain Julia. You can install Knet with the 
following at the julia prompt: `using Pkg; Pkg.add("Knet")`. Some starting points:

* [Tutorial:](tutorial) 
  introduces Julia and Knet via examples.
* [Documentation:](https://denizyuret.github.io/Knet.jl/latest)
  installation, introduction, design, implementation, full reference and deep learning chapters.
* [Examples:](examples)
  more tutorials and example models.
* [Benchmarks:](http://denizyuret.github.io/Knet.jl/latest/tutorial/#Benchmarks-1)
  comparison of Knet's speed with TensorFlow, PyTorch, DyNet etc.
* [Paper:](https://goo.gl/zeUBFr)
  Yuret, D. "Knet: beginning deep learning with 100 lines of julia." In *Machine Learning Systems Workshop* at NIPS 2016.
* [KnetML:](https://github.com/KnetML)
  github organization with Knet repos of models, tutorials, layer collections and other resources.
* [Images:](http://denizyuret.github.io/Knet.jl/latest/install/#Using-Amazon-AWS-1)
  Knet machine images are available for [AWS](http://denizyuret.github.io/Knet.jl/latest/install/#Using-Amazon-AWS-1), [Singularity](https://github.com/KnetML/singularity-images) and [Docker](https://github.com/JuliaGPU/docker).
* [Issues:](https://github.com/denizyuret/Knet.jl/issues)
  if you find a bug, please open a github issue.
* [knet-users:](https://groups.google.com/forum/#!forum/knet-users)
  if you need help or would like to request a feature, please join this mailing list.
* [knet-dev:](https://groups.google.com/forum/#!forum/knet-dev)
  if you would like to contribute to Knet development, please join this mailing list and check out these [tips](https://denizyuret.github.io/Knet.jl/latest/install/#Tips-for-developers-1).
* [knet-slack:](https://julialang.slack.com/messages/CDLKQ92P3/details) Slack channel for Knet.
* Related work: Please check out [Flux](https://github.com/FLuxML), [Mocha](https://github.com/pluskid/Mocha.jl), [JuliaML](https://github.com/JuliaML), [JuliaDiff](https://github.com/JuliaDiff), [JuliaGPU](https://github.com/JuliaGPU), [JuliaOpt](https://github.com/JuliaOpt) for related packages.

## Example

Here is a simple example where we define, train and test the
[LeNet](http://yann.lecun.com/exdb/lenet) model for the
[MNIST](http://yann.lecun.com/exdb/mnist) handwritten digit recognition dataset from scratch
using 15 lines of code and 10 seconds of GPU computation.

```julia
# Install packages before first run: using Pkg; pkg"add Knet IterTools MLDatasets"
using Knet, IterTools, MLDatasets

# Define convolutional layer:
struct Conv; w; b; end
Conv(w1,w2,nx,ny) = Conv(param(w1,w2,nx,ny), param0(1,1,ny,1))
(c::Conv)(x) = relu.(pool(conv4(c.w, x) .+ c.b))

# Define dense layer:
struct Dense; w; b; f; end
Dense(i,o; f=identity) = Dense(param(o,i), param0(o), f)
(d::Dense)(x) = d.f.(d.w * mat(x) .+ d.b)

# Define a chain of layers and a loss function:
struct Chain; layers; end
(c::Chain)(x) = (for l in c.layers; x = l(x); end; x)
(c::Chain)(x,y) = nll(c(x),y)

# Load MNIST data:
xtrn,ytrn = MNIST.traindata(Float32); ytrn[ytrn.==0] .= 10
xtst,ytst = MNIST.testdata(Float32);  ytst[ytst.==0] .= 10
dtrn = minibatch(xtrn, ytrn, 100; xsize = (28,28,1,:))
dtst = minibatch(xtst, ytst, 100; xsize = (28,28,1,:))

# Define and train LeNet (~10 secs on a GPU or ~3 mins on a CPU to reach ~99% accuracy)
LeNet = Chain((Conv(5,5,1,20), Conv(5,5,20,50), Dense(800,500,f=relu), Dense(500,10)))
progress!(adam(LeNet, ncycle(dtrn,3)))
accuracy(LeNet,data=dtst)
```

## Contributing

Knet is an open-source project and we are always open to new contributions: bug reports and
fixes, feature requests and contributions, new machine learning models and operators,
inspiring examples, benchmarking results are all welcome. See [Tips for Developers](https://denizyuret.github.io/Knet.jl/latest/install/#Tips-for-developers) for instructions.

Contributors: Can Gümeli, Carlo Lucibello, Ege Onat, Ekin Akyürek, Ekrem Emre Yurdakul, Emre Ünal, Emre Yolcu, Enis Berk, Erenay Dayanık, İlker Kesen, Kai Xu, Meriç Melike Softa, Mike Innes, Onur Kuru, Ozan Arkan Can, Ömer Kırnap, Phuoc Nguyen, Rene Donner, Tim Besard, Zhang Shiwei.


NEWS.md
Knet v1.4.10 Release Notes
=========================

* switch default GPU array type from KnetArray to CuArray
* allow loading on Apple M1
* support SpecialFunctions 2.x
* fix674: libknet8 compile error on unknown architectures
* tutorial updates
* minor test fixes


Knet v1.4.9 Release Notes
=========================
4720782 2021-10-13

* progress on layers21
* added vae chapter to docs
* updated libknet8 for windows
* made tutorial colab compatible


Knet v1.4.8 Release Notes
=========================
13fea96 2021-08-08

* fixed libknet8 path


Knet v1.4.7 Release Notes
=========================
d4af838 2021-07-23

* Compatibility fixes for Julia 1.6/1.7 and CUDA 3.x.
* Starting ops21, layers21, models21.


Knet v1.4.6 Release Notes
=========================
1492e2b 2021-03-17

* Improve fill! performance
* Fix #653: Knet.@save macro
* Fix #658 LinearIndices(::KnetArray)
* Use load->load143 for tutorials


Knet v1.4.5 Release Notes
=========================
5aff4be 2020-12-15

* Fixed #624: gcnode "key not found" bug.
* Fixed #647: Knet.save/load/@save/@load now call FileIO and JLD2 versions, use load143 to load old files.
* Removed warning about NNlib pooling bug that was fixed.


Knet v1.4.4 Release Notes
=========================
3b6f70a 2020-11-28

* Serialization and JLD2 support for KnetArray and RNN.
* Change eltype to Any in container types in serialize.jl.
* Compat fixes with CUDA 2.3 and Julia 1.6.
* Fixed #638 causing KnetArray broadcast/materialize!/dotview issue.
* Fixed Knet.seed! bug. (@egeonat)
* Added powerpc support. (@jdad)
* Fixed mnist labels in examples.


Knet v1.4.3 Release Notes
=========================
8a4fdbf 2020-10-16

* Upgrade to CUDA 2.0.
* Doc fixes.


Knet v1.4.2 Release Notes
=========================
701ecff 2020-09-28

* Fixed windows git-tree-sha1 for libknet8.
* Tutorial fixes.
* Fix 606: gcnode issues.
* Fix 610: WeakRefs turn into nothing when Julia garbage collects them.
* Fix 618: New GPUArrays indexing causes scalar indexing for some Knet operations (i.e., cat).
* Fix 619: Error converting CuArray to KnetArray.
* Fix 620: `k[1:2:3,:] .= 0` broken for KnetArray.


Knet v1.4.1 Release Notes
=========================
b720020 2020-08-28

* Tutorial, README fixes, using MLDatasets.
* Fixed gcnode issue #606.
* Use NVML when choosing GPU.
* Make RNN atype robust to types with dimensions.


Knet v1.4.0 Release Notes
=========================
2754cd6 2020-08-19

* Major refactoring of code without effecting the API (hopefully).
* CuArray support added to all ops, implemented gcnode, tested on examples and tutorial.
* Operators collected in Knet.Ops20: #583.
* Using @retry_reclaim on cudnn functions for stability.
* Fix #502 StackOverflowError when broadcast between number and KnetArray{Bool}.
* Started using MLDatasets in examples where possible.
* Profiling KnetArrays vs CuArrays in prof/ops.jl: #588.
* Removed deprecated directories.


Knet v1.3.9 Release Notes
=========================
1243060 2020-07-28

* Replaced all @cuda calls with CUDA.jl calls, leaving no dependence on external CUDA libraries except for the ones that come with CUDA.jl.
* Added libknet8 as an artifact for easier installation, removing the requirement to have nvcc and a host compiler.


Knet v1.3.8 Release Notes
=========================
2667e29 2020-07-24

* Update CUDA.jl version.
* CuArray performance improvements.


Knet v1.3.7 Release Notes
=========================
6256a7b 2020-07-12

* Fix #500: switch to using CUDA.jl instead of CuArrays, CUDAapi etc.
* Fix #571: travis doc deployment issue solved.
* GPU memory improvement: use weakrefs in gcnode.jl to prevent hanging on to old tapes.


Knet v1.3.6 Release Notes
=========================
9b30af3 2020-07-04

* Fix #562: conv performance improvements.
* doc fix and KnetArray method for nll, fixing #563.
* Fix #561: Loading Knet breaks Julia's copyto! on 1.5.
* Added update!(::Param,::Nothing) method to catch 0 gradient updates.
* Added bessel support to GPU.
* Further notes on GPU tools, especially for win (RocketRoss)
* Fix #558: progress and minimize get size(x,d...) methods to support collect with Julia 1.4.
* removed CUDAapi 1.0, 2.0 from Project.toml, not compat any more.
* removed julia 1.0 (fails) and 1.3 (redundant) from travis testing.
* added KnetArray(::CuArray) converter with shared pointer.
* windows test fixes and install doc updates.
* updated to work with CUDAapi v4.0 (iuliancioarca)
* add gpu install docs for azure/ubuntu18.04 (Jan Weidner)

Knet v1.3.5 Release Notes
=========================
4dd257a 2020-03-29

* CI fixes.
* cuda_visible_devices fix.
* Warn when trying KnetArray without GPU.
* Julia 1.4 compatibility fixes.


Knet v1.3.4 Release Notes
=========================
8c50f62 2020-02-29

* Tutorial notebook fixes.


Knet v1.3.3 Release Notes
=========================
239e838 2020-02-01

* Fixed bmm bug. (@ekinakyurek)
* Document fixes. (@Alexander-Barth)
* Broken conv tests now pass after NNLib fix.
* Tutorial fixes.
* RNN is robust against 0 input size.
* Progress is robust against empty iterators.
* Added special stop symbol to progress.
* Updated JuliaOnColab. (@ozanarkancan)
* Compatibility with new CUDA stack.


Knet v1.3.2 Release Notes
=========================
b386430 2019-11-29

* Compatibility with CUDAapi 2.0.
* Updated colab script. (@ozanarkancan)
* Fixed xavier = xavier_uniform and added xavier_normal distributions. (@Alexander-Barth)
* Conv now allocates workspace on demand and prefers algorithms with no workspace. (Issue #518)
* The progress bar function argument takes the Progress object as its argument instead of currval.


Knet v1.3.1 Release Notes
=========================
c94c1aa 2019-11-07

* Compatibility with CuArrays 1.3, 1.4.


Knet v1.3.0 Release Notes
=========================
55c01f4 2019-10-25

* Fixed #506: RNN serialization and gc issues.
* Fixed #509: Default output for progress is now stderr.
* Solved curand out of memory problem.

src/gpu.jl
using CUDA, TimerOutputs, Libdl, Pkg.Artifacts
const libknet8 = Libdl.find_library(["libknet8"], [artifact"libknet8"])
const tk = CUDA.find_toolkit()
const tkver = isempty(tk) ? v"0" : CUDA.parse_toolkit_version(CUDA.find_cuda_binary("ptxas", tk)) # v"0" because "nothing" will crash precompilation
const cudnnver = v"7" # cudnn version not read automatically. hardcoded to v7
const to = TimerOutput()
const Cptr = Ptr{Cvoid}
function getErrorString end

# moved profiling option from Knet.jl to gpu.jl to make it self contained for testing
const TIMER = haskey(ENV,"KNET_TIMER")

macro cudacall(lib,fun,returntype,argtypes,argvalues,errmsg=true,notfound=:(error("Cannot find $lib")))
    lib = string(lib); fun = string(fun)
	if isa(argtypes,Expr); argtypes = argtypes.args; end
    if isa(argvalues,Expr); argvalues = argvalues.args; end
	if lib == "cudnn"
		path = CUDA.find_cuda_library(lib,tk,cudnnver)
	elseif lib == "knet8"
		path = libknet8
	else
		path = CUDA.find_cuda_library(lib,tk,tkver)
	end
    if path==nothing || path==""; return notfound; end
    fx = Expr(:call, :ccall, Expr(:tuple,fun,path), returntype, Expr(:tuple,argtypes...), argvalues...)
    r = gensym()
    fx = Expr(:block,esc(:($r=$fx)))
    if errmsg || TIMER
        if TIMER
            push!(fx.args, esc(:(ccall(("cudaDeviceSynchronize","libcudart"),UInt32,()))))
        end
        if errmsg
            push!(fx.args, esc(:(if $r!=0; error($(@__MODULE__).getErrorString($lib,$fun,$r)); end)))
        else
            push!(fx.args, esc(r))
        end
        if TIMER
            fx = :(@timeit to $fun ($fx))
        end
    end
    esc(fx)
end

# macro cudnn(fun, argtypes, argvalues...); :(@cudacall("cudnn",$fun,UInt32,$argtypes,$argvalues)); end
# macro cuda(fun, argtypes, argvalues...); :(@cudacall("cuda",$fun,UInt32,$argtypes,$argvalues)); end
# macro cudart(fun, argtypes, argvalues...); :(@cudacall("cudart",$fun,UInt32,$argtypes,$argvalues)); end
# macro cudart1(fun, argtypes, argvalues...); :(@cudacall("cudart",$fun,UInt32,$argtypes,$argvalues,false,-1)); end # don't throw error
# macro cublas(fun, argtypes, argvalues...); :(@cudacall("cublas",$fun,UInt32,$argtypes,$argvalues)); end
# macro curand(fun, argtypes, argvalues...); :(@cudacall("curand",$fun,UInt32,$argtypes,$argvalues)); end
# macro nvml(fun, argtypes, argvalues...); :(@cudacall("nvml",$fun,UInt32,$argtypes,$argvalues)); end
macro knet8(fun, argtypes, argvalues...); :(@cudacall("knet8",$fun,Nothing,$argtypes,$argvalues,false)); end
macro knet8r(fun, returntype, argtypes, argvalues...); :(@cudacall("knet8",$fun,$returntype,$argtypes,$argvalues,false)); end # specify return type


"""

`gpu()` returns the id of the active GPU device or -1 if none are
active.

`gpu(true)` resets all GPU devices and activates the one with the most
available memory.

`gpu(false)` resets and deactivates all GPU devices.

`gpu(d::Int)` activates the GPU device `d` if `0 <= d < gpuCount()`,
otherwise deactivates devices.

`gpu(true/false)` resets all devices.  If there are any allocated
KnetArrays their pointers will be left dangling.  Thus
`gpu(true/false)` should only be used during startup.  If you want to
suspend GPU use temporarily, use `gpu(-1)`.

`gpu(d::Int)` does not reset the devices.  You can select a previous
device and find allocated memory preserved.  However trying to operate
on arrays of an inactive device will result in error.

"""
function gpu end

GPU = -1
gpu() = GPU
gpu(d::Integer) = (global GPU; GPU==d ? d : (try CUDA.device!(d); GPU=d; catch; GPU=-1; end))
gpu(b::Bool) = (global GPU=try b ? CUDA.device().handle : (CUDA.device_reset!();-1); catch; -1; end)

gpuCount() = length(CUDA.devices())
gpufree() = Mem.info()[1] + CUDA.pool[].cached_memory()

#=
let GPU=-1, GPUCNT=-1, CUBLAS=nothing, CUDNN=nothing, CURAND=nothing
    global gpu, gpuCount, cublashandle, cudnnhandle, curandGenerator

    gpu()=GPU

    function gpu(usegpu::Bool)
        global cudaRuntimeVersion, cudaDriverVersion, nvmlDriverVersion, nvmlVersion, nvmlfound, cudartfound
        ENV["CUDA_DEVICE_ORDER"]="PCI_BUS_ID" # From @ekinakyurek #541
        if !isdefined(@__MODULE__,:cudartfound)
            try #if (cudartfound = (Libdl.find_library(["libcudart"],[]) != ""))
                cudaRuntimeVersion = (p=Cint[0];@cudart(cudaRuntimeGetVersion,(Ptr{Cint},),p);Int(p[1]))
                cudaDriverVersion  = (p=Cint[0];@cudart(cudaDriverGetVersion, (Ptr{Cint},),p);Int(p[1]))
                cudartfound = true
            catch
                cudartfound = false
            end
        end
        if !isdefined(@__MODULE__,:nvmlfound)
            try #if (nvmlfound = (Libdl.find_library(["libnvidia-ml"],[]) != ""))
                # This code is only run once if successful, so nvmlInit here is ok
                @nvml(nvmlInit,())
                s = zeros(UInt8,80)
                @nvml(nvmlSystemGetDriverVersion,(Ptr{Cchar},Cuint),s,80)
                nvmlDriverVersion = unsafe_string(pointer(s))
                @nvml(nvmlSystemGetNVMLVersion,(Ptr{Cchar},Cuint),s,80)
                nvmlVersion = unsafe_string(pointer(s))
                # Let us keep nvml initialized for future ops such as meminfo
                # @nvml(nvmlShutdown,())
                nvmlfound = true
            catch
                nvmlfound = false
            end
        end
        if usegpu && gpuCount() > 0
            if GPU >= 0
                return
            elseif gpuCount() == 1
                pick = 0
            else # Pick based on memory usage
                pick = free = same = -1
                for i=0:gpuCount()-1
                    if nvmlfound
                        freemem = nvmlDeviceGetMemoryInfo(nvmlid(i))[2]
                    else
                        @cudart(cudaSetDevice, (Cint,), i)
                        freemem = cudaMemGetInfo()[1]
                        @cudart(cudaDeviceReset,()) # otherwise we leave a memory footprint
                    end
                    if freemem > free
                        pick = i
                        free = freemem
                        same = 1
                    elseif freemem == free
                        # pick one of equal devices randomly
                        rand(1:(same+=1)) == 1 && (pick = i)
                    end
                end
            end
            @cudart(cudaDeviceReset,())  # #541: clear memory
            gpu(pick)
        elseif gpuCount() > 0
            @cudart(cudaDeviceReset,())  # #541: clear memory
            gpu(-1)
        else
            gpu(-1)
        end
    end

    function gpu(i::Int)
        if GPU == i
            # nothing to do
        elseif 0 <= i < gpuCount()
            GPU = i
            @cudart(cudaSetDevice, (Cint,), i)

            # Initializing CUDAnative helps with stability problems for some devices
            # However cuda, nvml and cu use different device numbers! (i) is the cuda device number, CUDAdrv uses cu numbers
            # We find the equivalent cu number from pciBusId:
            # https://stackoverflow.com/questions/13781738/how-does-cuda-assign-device-ids-to-gpus
            # 20200108: CUDAdrv 5.0 initializes to currently active device, so this is no longer needed.
            # CUDAnative.initialize(CUDAdrv.CuDevice(cuid(i)))
            # 20200707: @maleadt added this:
            CUDA.device!(cuid(i))

            # Initialize curand to guard against gpu memory fillup before first dropout (#181)
            curandGenerator()
            
        else
            GPU = -1
            # @cudart(cudaDeviceReset,()) # may still go back and use arrays allocated in a previous gpu
        end
        return GPU
    end

    function gpuCount() # should not bomb when there is no gpu or nvidia libs
        if GPUCNT == -1
            GPUCNT = try
	        p=Cuint[0]
                # if nvmlfound
                # # We prefer nvml because cudart takes up memory even if we don't use a device
                #   @nvml(nvmlDeviceGetCount,(Ptr{Cuint},),p)
                # #541: nvml does not respect visible devices (@ekinakyurek)
                # also OSX does not have the nvidia-ml library!
                if cudartfound
                    @cudart1(cudaGetDeviceCount,(Ptr{Cuint},),p)
                end
	        Int(p[1])
            catch
	        0
            end
        end
        return GPUCNT
    end

    function cublashandle(dev=gpu())
        dev==-1 && return nothing
        i = dev+2
        if CUBLAS === nothing
            CUBLAS=Array{Any}(nothing,gpuCount()+1)
        end
        if CUBLAS[i] === nothing
            @assert dev == gpu()
            CUBLAS[i]=cublasCreate()
        end
        return CUBLAS[i]
    end

    function cudnnhandle(dev=gpu())
        dev==-1 && return nothing
        i = dev+2
        if CUDNN === nothing
            CUDNN=Array{Any}(nothing,gpuCount()+1)
        end
        if CUDNN[i] === nothing
            @assert dev == gpu()
            CUDNN[i]=cudnnCreate()
        end
        return CUDNN[i]
    end

    function curandGenerator(dev=gpu(); seed=nothing)
        dev==-1 && return nothing
        i = dev+2
        if CURAND === nothing
            CURAND=Array{Any}(nothing,gpuCount()+1)
        end
        if CURAND[i] === nothing || seed !== nothing
            @assert dev == gpu()
            if CURAND[i] !== nothing
                @curand(curandDestroyGenerator,(Cptr,),CURAND[i])
                CURAND[i] = nothing  # to prevent double free if curandCreateGenerator errors
            end
            CURAND[i]=curandCreateGenerator(seed)
        end
        return CURAND[i]
    end
end

# cudaGetDeviceCount is deprecated, use gpuCount instead:
cudaGetDeviceCount()=(p=Cuint[0];@cudart1(cudaGetDeviceCount,(Ptr{Cuint},),p);Int(p[1])) # will not bomb when there is no gpu
cudaGetDevice()=(d=Cint[-1];@cudart(cudaGetDevice,(Ptr{Cint},),d);Int(d[1]))
"Returns free,total memory."
cudaMemGetInfo()=(f=Csize_t[0];m=Csize_t[0]; @cudart(cudaMemGetInfo,(Ptr{Csize_t},Ptr{Csize_t}),f,m); (Int(f[1]),Int(m[1])))
cudaDeviceSynchronize()=@cudart(cudaDeviceSynchronize,())

# cuda, cudart, nvml each number devices differntly
# we are using the cudart ids, here is some converters:

function cudaDeviceGetPCIBusId(i=gpu())
    pciBusId = zeros(UInt8,16)
    @cudart(cudaDeviceGetPCIBusId, (Cptr, Cint, Cint), pciBusId, length(pciBusId), i)
    return unsafe_string(pointer(pciBusId))
end

function cuid(i=gpu())
    pci = cudaDeviceGetPCIBusId(i)
    id = Cint[0]
    @cuda(cuDeviceGetByPCIBusId, (Cptr, Cptr), id, pointer(pci))
    return Int(id[1])
end

function nvmlid(i=gpu())
    pci = cudaDeviceGetPCIBusId(i)
    hnd = Cptr[0]; idx = Cuint[0];
    @nvml(nvmlDeviceGetHandleByPciBusId, (Cptr, Cptr), pointer(pci), hnd)
    @nvml(nvmlDeviceGetIndex, (Cptr, Cptr), hnd[1], idx)
    return Int(idx[1])
end


"Returns total,free,used memory."
function nvmlDeviceGetMemoryInfo(i=nvmlid(gpu()))
    0 <= i || return nothing
    dev = Cptr[0]
    mem = Array{Culonglong}(undef,3)
    @nvml("nvmlDeviceGetHandleByIndex",(Cuint,Ptr{Cptr}),i,dev)
    @nvml("nvmlDeviceGetMemoryInfo",(Cptr,Ptr{Culonglong}),dev[1],mem)
    ntuple(i->Int(mem[i]),length(mem))
end

function gpumem(i=gpu())
    if i < 0 || i >= gpuCount()
        return (0,0,0)
    elseif nvmlfound
        return nvmlDeviceGetMemoryInfo(nvmlid(i))
    elseif cudartfound
        dev=gpu(); gpu(i)
        free,total = cudaMemGetInfo()
        gpu(dev)
        return (total,free,total-free)
    else
        error("gpumem cannot find nvml or cudart.")
    end
end

gpufree(i=gpu())=gpumem(i)[2]

function cublasCreate()
    handleP = Cptr[0]
    @cublas(cublasCreate_v2, (Ptr{Cptr},), handleP)
    handle = handleP[1]
    atexit(()->@cublas(cublasDestroy_v2, (Cptr,), handle))
    global cublasVersion = (p=Cint[0];@cublas(cublasGetVersion_v2,(Cptr,Ptr{Cint}),handle,p);Int(p[1]))
    return handle
end

function cudnnCreate()
    path = CUDA.find_cuda_library("cudnn",tk,cudnnver)
    if path==nothing; error("Cannot find cudnn"); end
    handleP = Cptr[0]
    @cudnn(cudnnCreate,(Ptr{Cptr},), handleP)
    handle = handleP[1]
    atexit(()->@cudnn(cudnnDestroy,(Cptr,), handle))
    global cudnnVersion = Int(eval(:(ccall(("cudnnGetVersion",$path),Csize_t,()))))
    return handle
end

function curandCreateGenerator(seed = nothing)
    CURAND_RNG_PSEUDO_DEFAULT = 100 # Default pseudorandom generator
    r = Cptr[0]
    @curand(curandCreateGenerator,(Cptr,Cint),r,CURAND_RNG_PSEUDO_DEFAULT)
    seed !== nothing && @curand(curandSetPseudoRandomGeneratorSeed,(Cptr,Culonglong),r[1],seed)
    # This call ensures memory buffer allocation. Without it we get "out of memory" later:
    @curand(curandGenerateUniform,(Cptr,Ptr{Cfloat},Csize_t),r[1],C_NULL,0)
    return r[1]
end

const curanderrors = Dict(
    0 => "No errors",
    100 => "Header file and linked library version do not match",
    101 => "Generator not initialized",
    102 => "Memory allocation failed",
    103 => "Generator is wrong type",
    104 => "Argument out of range",
    105 => "Length requested is not a multiple of dimension",
    106 => "GPU does not have double precision required by MRG32k3a",
    201 => "Kernel launch failure",
    202 => "Preexisting failure on library entry",
    203 => "Initialization of CUDA failed",
    204 => "Architecture mismatch, GPU does not support requested feature",
    999 => "Internal library error",
)

const cublaserrors = Dict(
    0 => "CUBLAS_STATUS_SUCCESS",
    1 => "CUBLAS_STATUS_NOT_INITIALIZED",
    3 => "CUBLAS_STATUS_ALLOC_FAILED",
    7 => "CUBLAS_STATUS_INVALID_VALUE",
    8 => "CUBLAS_STATUS_ARCH_MISMATCH",
    11 => "CUBLAS_STATUS_MAPPING_ERROR",
    13 => "CUBLAS_STATUS_EXECUTION_FAILED",
    14 => "CUBLAS_STATUS_INTERNAL_ERROR",
    15 => "CUBLAS_STATUS_NOT_SUPPORTED",
    16 => "CUBLAS_STATUS_LICENSE_ERROR",
)

function getErrorString(lib,fun,ret)
    if lib == "cudnn"
	path = CUDA.find_cuda_library(lib,tk,cudnnver)
    else
	path = CUDA.find_cuda_library(lib,tk,tkver)
    end
    if lib == "cudart" && path != nothing
        str = unsafe_string(@eval(ccall(("cudaGetErrorString",$path),Cstring,(UInt32,),$ret)))
    elseif lib == "cudnn" && path != nothing
        str = unsafe_string(@eval(ccall(("cudnnGetErrorString",$path),Cstring,(UInt32,),$ret)))
    elseif lib == "nvml" && path != nothing
        str = unsafe_string(@eval(ccall(("nvmlErrorString",$path),Cstring,(UInt32,),$ret)))
    elseif lib == "curand"
        str = get(curanderrors,ret,"Unknown $lib error in $fun")
    elseif lib == "cublas"
        str = get(cublaserrors,ret,"Unknown $lib error in $fun")
    else
        str = "Unknown $lib error in $fun"
    end
    string(fun, ": ", ret, ": ", str)
end

=#
